File: merge_sort.py
# 2.0503997802734375e-05 for 10
# 0.00011467933654785156 for 100
# 0.00135970115661621100 for 1000
# 0.01644515991210937500 for 10000
# 0.21151900291442870000 for 100000
# 2.61152029037475600000 for 1000000
def merge_sort(values):
    if len(values) == 2 and values[0] > values[1]:
        values[0], values[1] = values[1], values[0]
    elif len(values) > 2:
        b, c = values[:len(values) // 2], values[len(values) // 2:]
        merge_sort(b)
        merge_sort(c)
        i, j, k = (0, 0, 0)
        while i < len(b) and j < len(c):
            if b[i] < c[j]:
                values[k] = b[i]
                i += 1
            else:
                values[k] = c[j]
                j += 1
            k += 1
        while i < len(b):
            values[k] = b[i]
            i += 1
            k += 1
        while j < len(c):
            values[k] = c[j]
            j += 1
            k += 1

File: MergeSort.java
// 10: 5.093E-6
// 100: 3.6858E-5
// 1000: 4.31988E-4
// 10000: 0.001746159
// 100000: 0.055399026
// 1000000: 0.089316221
// 10000000: 0.839243773
// 100000000: 7.356650992
public class MergeSort {
    public static double[] sort(double[] values){
        if(values.length == 2 && values[0] > values[1]){
            double tmp = values[0];
            values[0] = values[1];
            values[1] = tmp;
        }else if(values.length > 2){
            double[] values2 = new double[values.length / 2];
            double[] values3 = new double[values.length - values2.length];
            for(int i = 0;i < values2.length;i++){
                values2[i] = values[i];
            }
            for(int i = 0;i < values3.length;i++){
                values3[i] = values[values2.length + i];
            }
            sort(values2);
            sort(values3);
            int i = 0, j = 0, k = 0;
            while(i < values2.length && j < values3.length){
                if(values2[i] < values3[j]){
                    values[k++] = values2[i++];
                }else{
                    values[k++] = values3[j++];
                }
            }
            while(i < values2.length){
                values[k++] = values2[i++];
            }
            while(j < values3.length){
                values[k++] = values3[i++];
            }
        }
        return values;
    }

    public static void test(int length){
        long start = System.nanoTime();
        double[] values = new double[length];
        for(int i = 0;i < length;i++){
            values[i] = length - i;
        }
        sort(values);
        long performance = System.nanoTime() - start;
        for(double value : values){
            System.out.println(value);
        }
        System.out.println(performance / 1000000000.0);
    }
}
